experiments/process_sandbox/filetree.h

```diff
@@ -0,0 +1,319 @@
+   * structure where each leaf node is either a directory handle to a real
+   * directory or a file handle to a file (or something equivalent, such as
+   * an anonymous shared memory object).
+   *
+   * Note: The directories in this structure that are not represented by
+   * handles cannot be opened or otherwise inspected in the sandbox.  If the
+   * tree stores `/tmp/example/somefile`, where only `somefile` is
+   * represented by a real file descriptor, the sandbox will not see `/tmp`
+   * or `/tmp/example`.  This restriction may be lifted in the future if it
+   * causes significant problems.
+   *
+   * TODO: The current implementation provides all-or-nothing permissions.
+   * Files or directories are delegated with ambient authority.  This is fine
+   * for simple cases but eventually we should provide finer-grained
+   * controls.  For example, when running a Verona program as `root`,
+   * sandboxed code should not have read-write access to `/lib`, even if the
+   * `root` user does.
+   */
+  class ExportedFileTree
+  {
+    /**
+     * A single directory within the exported tree.  This class exists mostly
+     * because C++ does not allow recursive types without an explicitly named
+     * class somewere in the cycle, so we can't define a map whose value type
+     * is the map type itself.
+     */
+    class ExportedDirectory
+    {
+      /**
+       * This class is an implementation detail of `ExportedFileTree`, which
+       * is the only thing allowed to interact with it directly.
+       */
+      friend class ExportedFileTree;
+
+      /**
+       * Handle representing a file.  This is used so that we can statically
+       * differentiate between handles to files and handles to directories in
+       * the type system but is in no other way different from a generic handle.
+       */
+      struct FileHandle : public platform::Handle
+      {
+        using platform::Handle::Handle;
+        FileHandle(Handle&& h) : Handle(std::move(h)) {}
+      };
+
+      /**
+       * Handle representing a directory.  This is used so that we can
+       * statically differentiate between handles to files and handles to
+       * directories in the type system but is in no other way different from a
+       * generic handle.
+       */
+      struct DirHandle : public platform::Handle
+      {
+        using platform::Handle::Handle;
+        DirHandle(Handle&& h) : Handle(std::move(h)) {}
+      };
+
+      /**
+       * A pointer to the child directory.  This could probably be a unique
+       * pointer as it's currently used, but `std::variant` can't contain
+       * references and so we'd have to use bare pointers during traversal.
+       */
+      using DirPtr = std::shared_ptr<ExportedDirectory>;
+
+      /**
+       * The contents of this virtual directory.  This maps from file names to
+       * another instance of this class or a file / directory handle.
```

Comment:
So, `DirPtr` is an object that I can do file lookup, while `DirHandle` / `FileHandle` are handlers to the actual dir/file objects in the file system?

This seems like a weird system to query file systems. Why not a simple dir/file handlers and a collection of methods (in a static class or something) that can traverse the actual file system?

Maybe this is a `view` into the filesystem that doesn't allow the sandbox to see everything, but only traverse certain directories and open certain files?